x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/FrozenCacheService.java

```diff
@@ -754,89 +681,133 @@ protected void alreadyClosed() {
+         * {@code rangeToRead}.
+         * NOTE: this method will only invoke the provided {@code writer} once with the maximum range necessary to populate the cache
+         * with the {@code rangeToWrite}. This is done with the assumption that each invocation of {@code writer} has a non-trivial
+         * overhead due e.g. sending separate requests to a blob-store that outweighs the cost of reading a some bytes that will simply
+         * be skipped because they already are in the cache.
+         *
+         * @param rangeToWrite range to populate cache with
+         * @param rangeToRead  range to pass to {@code reader} once available
+         * @param reader       reader to consume read bytes
+         * @param writer       writer that populates cache with requested {@code rangeToWrite}
+         * @param executor     executor to run IO operations on
+         * @return             listener that resolves with the number of bytes read
+         */
         public StepListener<Integer> populateAndRead(
             final ByteRange rangeToWrite,
             final ByteRange rangeToRead,
             final RangeAvailableHandler reader,
             final RangeMissingHandler writer,
             final Executor executor
         ) {
+            assert rangeToWrite.length() > 0 : "should not try to populate and read empty range";
             StepListener<Integer> stepListener = null;
-            final long writeStart = rangeToWrite.start();
-            final long readStart = rangeToRead.start();
-            for (int region = getRegion(rangeToWrite.start()); region <= getEndingRegion(rangeToWrite.end()); region++) {
-                final ByteRange subRangeToWrite = mapSubRangeToRegion(rangeToWrite, region);
-                final ByteRange subRangeToRead = mapSubRangeToRegion(rangeToRead, region);
-                if (subRangeToRead.length() == 0L) {
-                    // nothing to read, skip
-                    if (stepListener == null) {
-                        stepListener = new StepListener<>();
-                        stepListener.onResponse(0);
-                    }
-                    continue;
+            final int lastRegion = regionForPosition(rangeToWrite.end() - 1);
+            final int firstRegion = regionForPosition(rangeToWrite.start());
+            final PopulateAndReadContext context = new PopulateAndReadContext();
+            for (int region = firstRegion; region <= lastRegion; region++) {
+                final CacheFileRegion fileRegion;
+                try {
+                    fileRegion = get(cacheKey, fileSize, region);
+                } catch (Exception e) {
+                    // if we can't acquire enough cache pages for all the regions in a write we fail the execution right away so that the
+                    // caller can move on to a direct read bypassing the cache
+                    context.fail(e);
```

Comment:
as discussed, I think there's still value in not aborting here if the requested region is outside of the read range.